core/services.go

```diff
@@ -0,0 +1,348 @@
+	case isRunning:
+		ss.l.Unlock()
+		return running, nil
+	case isStarting:
+		ss.l.Unlock()
+		starting.Wait()
+		ss.l.Lock()
+		running, isRunning = ss.running[key]
+		ss.l.Unlock()
+		if isRunning {
+			return running, nil
+		}
+		return nil, notRunningErr
+	default:
+		return nil, fmt.Errorf("internal error: unexpected state")
+	}
+}
+
+// StartBindings starts each of the bound services in parallel and returns a
+// function that will detach from all of them after 10 seconds.
+func (ss *Services) StartBindings(ctx context.Context, bk *buildkit.Client, bindings ServiceBindings) (_ func(), _ []*RunningService, err error) {
+	running := []*RunningService{}
+	detach := func() {
+		go func() {
+			<-time.After(DetachGracePeriod)
+			for _, svc := range running {
+				ss.Detach(ctx, svc)
+			}
+		}()
+	}
+
+	defer func() {
+		if err != nil {
+			detach()
+		}
+	}()
+
+	// NB: don't use errgroup.WithCancel; we don't want to cancel on Wait
+	eg := new(errgroup.Group)
+
+	started := make(chan *RunningService, len(bindings))
+	for _, bnd := range bindings {
+		bnd := bnd
+		eg.Go(func() error {
+			runningSvc, err := ss.Start(ctx, bnd.Service)
+			if err != nil {
+				return fmt.Errorf("start %s (%s): %w", bnd.Hostname, bnd.Aliases, err)
+			}
+			started <- runningSvc
+			return nil
+		})
+	}
+
+	startErr := eg.Wait()
+
+	close(started)
+
+	if startErr != nil {
+		return nil, nil, startErr
+	}
+
+	for svc := range started {
+		running = append(running, svc)
+	}
+
+	return detach, running, nil
+}
+
+// Start starts the given service, returning the running service. If the
+// service is already running, it is returned immediately. If the service is
+// already starting, it waits for it to finish and returns the running service.
+// If the service failed to start, it tries again.
+func (ss *Services) Start(ctx context.Context, svc *Service) (*RunningService, error) {
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	dig, err := svc.Digest()
+	if err != nil {
+		return nil, err
+	}
+
+	key := ServiceKey{
+		Digest:   dig,
+		ClientID: clientMetadata.ClientID,
+	}
+
+	ss.l.Lock()
+	starting, isStarting := ss.starting[key]
+	running, isRunning := ss.running[key]
+	switch {
+	case !isStarting && !isRunning:
+		// not starting or running; start it
+		starting = new(sync.WaitGroup)
+		starting.Add(1)
+		defer starting.Done()
+		ss.starting[key] = starting
+	case isRunning:
+		// already running; increment binding count and return
+		ss.bindings[key]++
+		ss.l.Unlock()
+		return running, nil
+	case isStarting:
+		// already starting; wait for the attempt to finish and check if it
+		// succeeded
+		ss.l.Unlock()
+		starting.Wait()
+		ss.l.Lock()
+		running, didStart := ss.running[key]
+		if didStart {
+			// starting succeeded as normal; return the isntance
+			ss.l.Unlock()
+			return running, nil
+		}
+		// starting didn't work; give it another go (this might just error again)
```

Comment:
Possible synchronization bug:
1. Multiple goroutines are blocked on `starting.Wait()` on line 232
2. Starting fails, so those goroutines all unblock and continue to `svc.Start` on line 250
3. I think that may already be unexpected behavior, but also `ss.bindings` will always be set to `1` on 262

Not sure if I'm just misunderstanding since it's all pretty tricky obviously.

Reply:
Yeah this was definitely sus. Pushed a theoretical fix, but I'll admit I haven't reproduced this first. Unit tests are in order, if I can isolate the concurrency aspect. :thinking: